compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,211 @@
+            extensionReceiverParameter = function.extensionReceiverParameter?.copyWithTypeErasure(this, typeMap)
+            function.valueParameters.mapIndexed { i, param ->
+                valueParameters.add(i, param.copyWithTypeErasure(this, typeMap))
+            }
+            val exception = context.getTopLevelClass(FqName("java.lang.UnsupportedOperationException"))
+            // Function body consist only of throwing UnsupportedOperationException statement
+            body = context.createIrBuilder(function.symbol).irBlockBody {
+                +irThrow(
+                    irCall(
+                        exception.owner.constructors.single {
+                            it.valueParameters.size == 1 && it.valueParameters.single().type.isNullableString()
+                        }
+                    ).apply {
+                        putValueArgument(0, irString("Operation is not supported for read-only collection"))
+                    }
+                )
+            }
+        }
+    }
+
+    // Type erasure for value parameters, with type substitution
+    private fun IrValueParameter.copyWithTypeErasure(target: IrSimpleFunction, typeMap: Map<IrTypeParameterSymbol, IrType>)
+            : IrValueParameter {
+        val descriptor = WrappedValueParameterDescriptor(this.descriptor.annotations)
+        return IrValueParameterImpl(
+            UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+            IrDeclarationOrigin.IR_BUILTINS_STUB,
+            IrValueParameterSymbolImpl(descriptor),
+            name,
+            index,
+            type.substitute(typeMap).eraseTypeParameters(),
+            varargElementType?.substitute(typeMap)?.eraseTypeParameters(),
+            isCrossinline,
+            isNoinline
+        ).apply {
+            descriptor.bind(this)
+            parent = target
+        }
+    }
+
+    // Create a map for type parameters between source class (Mutable Collection classes) to
+    // target class (class currently in lowering phase), this map is later used for substituting type parameters in generated functions
+    private fun createTypeMap(readOnlyClass: IrClass, mutableClass: IrClass, targetClass: IrClass): Map<IrTypeParameterSymbol, IrType> {
+        // We find the most specific type for the immutable collection class from the inheritance chain of target class
+        // Perform type substitution along searching, then use the type arguments obtained from the most specific type
+        // for type substitution.
+        val readOnlyClassType = getAllSuperTypes(targetClass.defaultType).findMostSpecificTypeForClass(readOnlyClass.symbol)
+        val targetTypeList = (readOnlyClassType as IrSimpleType).arguments.mapNotNull { (it as IrTypeProjection).type }
+
+        if (targetTypeList.isEmpty() || targetTypeList.size != mutableClass.typeParameters.size) {
+            throw IllegalStateException("Type argument mismatch between immutable and mutable class")
```

Comment:
Please always add context to exception messages. For example, at least `targetClass.fqNameWhenAvailable` here, but I'd also prefer to render some information about either `readOnlyClass` or `mutableClass`, and `targetTypeList`